### PR TITLE
Adding clipping plane functionality

### PR DESCRIPTION
**Pull request by nashmit**
Implementing clipping plane functionality:

- adapting changes to the Vertex and Geometry shader to accommodate custom clipping planes.
- implemented an automatic management system for OpenGL clipping planes.

Currently, the number of clipping planes actively supported in parallel by OpenGL is 8. 
All planes are defined in the global reference frame.

### CODE REVIEW

File: gui/widgets/qopenglmousetracker.cpp [hunk 1]
```diff
@@ -81,6 +81,17 @@ void QOpenGLMouseTracker::renderScene(std::optional<float> lineWidth) {
   }
 }
 
+void QOpenGLMouseTracker::updateAllClippingPlanes(
+    std::unique_ptr<rendering::ShaderProgram> &program) const {
```
Comment (lkeegan) [inline gui/widgets/qopenglmousetracker.cpp:85]: does this need to take program as an argument - isn't it always going to use the member variable program?

File: gui/rendering/ClippingPlane.cpp [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  SetClipPlane(normal.x(), normal.y(), normal.z(),
+               -QVector3D::dotProduct(normal, point));
+}
+
+void rendering::ClippingPlane::TranslateClipPlane(GLfloat value) {
+
+  GLfloat length = QVector3D(a, b, c).length();
+  QVector3D normal(a / length, b / length, c / length);
+
+  GLfloat p = d / length;
```
Comment (lkeegan) [inline gui/rendering/ClippingPlane.cpp:87]: isn't length=1 here?

**Review by lkeegan — COMMENTED** [review]
Thanks, a comment about the interface and ownership:

- Interface: up to now the external interface of the QOpenGLWidget didn't expose the rendering objects
  - Now `ClippingPlane` is part of its API
  - It would be nice not to expose this implementation detail
- Ownership: nothing owns these ClippingPlane objects
  - The natural owner to me would be the QOpenGLWidget?

Also, since the number of planes is fixed, it might be simpler to just have a std::array of them in the QOpenGLWidget, and then enable/disable/modify them them instead of creating and deleting objects & maintaining a shared stack of available indices?

File: gui/rendering/ClippingPlane.cpp (hunk 2 above)
Comment (nashmit) [reply gui/rendering/ClippingPlane.cpp:87]: you can have normals of different size than 1.

Comment (nashmit) [reply gui/rendering/ClippingPlane.cpp:87]: No necessary... you can have valid planes like this:  (a,b,c,d) = (6,7,8,1)

File: gui/widgets/qopenglmousetracker.cpp [hunk 3]
```diff
@@ -20,6 +20,66 @@ QOpenGLMouseTracker::QOpenGLMouseTracker(QWidget *parent, float lineWidth,
       m_backgroundColor(QWidget::palette().color(QWidget::backgroundRole())),
       m_lastColour(QWidget::palette().color(QWidget::backgroundRole()).rgb()) {}
 
+std::shared_ptr<rendering::ClippingPlane>
+QOpenGLMouseTracker::BuildClippingPlane(GLfloat a, GLfloat b, GLfloat c,
+                                        GLfloat d, bool active) {
+
+  auto it = m_clippingPlanesPool.begin();
+  assert(it != m_clippingPlanesPool.end());
```
Comment (lkeegan) [inline gui/widgets/qopenglmousetracker.cpp:28]: if this could happen then it should be a run-time check - in a release build the assert does nothing and if the pool is empty the next line segfaults

File: gui/widgets/qopenglmousetracker.cpp [hunk 4]
```diff
@@ -20,6 +20,66 @@ QOpenGLMouseTracker::QOpenGLMouseTracker(QWidget *parent, float lineWidth,
...
+}
+
+std::shared_ptr<rendering::ClippingPlane>
+QOpenGLMouseTracker::BuildClippingPlane(const QVector3D &normal,
+                                        const QVector3D &point, bool active) {
+
+  auto it = m_clippingPlanesPool.begin();
+  assert(it != m_clippingPlanesPool.end());
+
+  auto clippingPlane = *it;
```
Comment (lkeegan) [inline gui/widgets/qopenglmousetracker.cpp:54]: again this iterator should be checked that it isn't `m_clippingPlanesPool.end()` before dereferencing it

File: gui/widgets/qopenglmousetracker.cpp [hunk 5]
```diff
@@ -20,6 +20,66 @@ QOpenGLMouseTracker::QOpenGLMouseTracker(QWidget *parent, float lineWidth,
...
+  update();
+
+  return clippingPlane;
+}
+
+void QOpenGLMouseTracker::DestroyClippingPlane(
+    std::shared_ptr<rendering::ClippingPlane> clippingPlane) {
+
+  auto it = m_clippingPlanes.find(clippingPlane);
+  assert(it != m_clippingPlanes.end());
```
Comment (lkeegan) [inline gui/widgets/qopenglmousetracker.cpp:75]: same here

File: gui/widgets/qopenglmousetracker.cpp [hunk 6]
```diff
@@ -20,6 +20,66 @@ QOpenGLMouseTracker::QOpenGLMouseTracker(QWidget *parent, float lineWidth,
       m_backgroundColor(QWidget::palette().color(QWidget::backgroundRole())),
       m_lastColour(QWidget::palette().color(QWidget::backgroundRole()).rgb()) {}
 
+std::shared_ptr<rendering::ClippingPlane>
+QOpenGLMouseTracker::BuildClippingPlane(GLfloat a, GLfloat b, GLfloat c,
```
Comment (lkeegan) [inline gui/widgets/qopenglmousetracker.cpp:24]: There are two QOpenGLMouseTracker::BuildClippingPlane functions that have the same contents except for one line, and the first one isn't tested - if not needed could you remove the unused code?
Or if you need both then could you refactor them to remove the copy&pasted logic?

**Review by lkeegan — COMMENTED** [review]
It would be good to add a couple of tests, e.g. DestroyClippingPlane is not tested at all as far as I can see, and trying to add more than 8 planes is a test I would write to check nothing goes wrong when the pool is empty

File: gui/widgets/qopenglmousetracker.cpp (hunk 6 above)
Comment (nashmit) [reply gui/widgets/qopenglmousetracker.cpp:24]: Both of them might be of use and are rapping other 2 functions that can be used for updating the planes. I don't have a nice way of getting away from the redundancy without adding extra functions that will also be unuseful for other than this case.

Comment (lkeegan) [reply gui/widgets/qopenglmousetracker.cpp:24]: OK, then I guess let's merge this as is and then see what functions end up being used in the full implementation

**Review by lkeegan — APPROVED** [review]
